src/aarch64/interrupt.c

```diff
@@ -0,0 +1,361 @@
+#include <kernel.h>
+#include <page.h>
+#include <symtab.h>
+#include <gic.h>
+
+//#define INT_DEBUG
+#ifdef INT_DEBUG
+#define int_debug(x, ...) do {log_printf("  INT", x, ##__VA_ARGS__);} while(0)
+#else
+#define int_debug(x, ...)
+#endif
+
+#define INTERRUPT_VECTOR_START 32 /* end of exceptions; defined by architecture */
+#define MAX_INTERRUPT_VECTORS  256 /* as defined by architecture; we may have less */
+
+typedef struct inthandler {
+    struct list l;
+    thunk t;
+    const char *name;
+} *inthandler;
+
+static struct list *handlers;
+
+extern u32 n_interrupt_vectors;
+extern u32 interrupt_vector_size;
+extern void * interrupt_vectors;
+
+static char* gpreg_names[FRAME_N_GPREG] = {
+    "  x0", "  x1", "  x2", "  x3", "  x4", "  x5", "  x6", "  x7",
+    "  x8", "  x9", " x10", " x11", " x12", " x13", " x14", " x15",
+    " x16", " x17", " x18", " x19", " x20", " x21", " x22", " x23",
+    " x24", " x25", " x26", " x27", " x28", " x29", " x30", "  sp" };
+
+static char* fpsimd_names[35] = {
+    "  q0", "  q1", "  q2", "  q3", "  q4", "  q5", "  q6", "  q7",
+    "  q8", "  q9", " q10", " q11", " q12", " q13", " q14", " q15",
+    " q16", " q17", " q18", " q19", " q20", " q21", " q22", " q23",
+    " q24", " q25", " q26", " q27", " q28", " q29", " q30", " q31",
+    "fpsr", "fpcr"};
+
+void install_fallback_fault_handler(fault_handler h)
+{
+    // XXX reconstruct
+    for (int i = 0; i < MAX_CPUS; i++) {
+        cpuinfo_from_id(i)->kernel_context->frame[FRAME_FAULT_HANDLER] = u64_from_pointer(h);
+    }
+}
+
+static void print_far_if_valid(u32 iss)
+{
+    if ((iss & ESR_ISS_DATA_ABRT_FnV) == 0) {
+        register u64 far;
+        asm("mrs %0, FAR_EL1" : "=r"(far));
+        console("\n       far: ");
+        print_u64_with_sym(far);
+    }
+}
+
+void print_frame(context f)
+{
+    u64 v = f[FRAME_VECTOR];
+    console(" interrupt: ");
+    print_u64(v);
```

Comment:
Ok. It looks like there are still a lot of calls to console, even in x86_64 code (like interrupt.c), so perhaps those should be changed over too.

What is the purpose of having both console() and rputs()? There could just be one call that does the console_write, and the klog_write can just be a no-op if klog isn't initialized.

Reply:
The remaining console() calls in x86_64/interrupt.c are there because when they happen the VM is about to be shut down via a direct vm_exit() call, so anything that we write to the log isn't going to be flushed to disk anyway.

I created an rputs() function separated from console() because the kernel log seemed to me conceptually different from a console, but I don't have a strong opinion on whether they should be unified.

Reply:
That's fine. I left the existing calls to console in place and changed those in src/aarch64 that should be logged to klog.